1\. **Pandas DataFrame**

This exercise consists in analyzing a dataset containg timing information from a series of Time-to-Digital-Converters (TDC) implemented in a pair of FPGAs. Each measurement (i.e. each row of the input file) consists of a flag that specifies the type of message ('HEAD', which in this case is always 1), two addresses of the TDC providing the signal ('FPGA' and 'TDC_CHANNEL'), and the timing information ('ORBIT_CNT', 'BX_COUNTER', and 'TDC_MEAS'). Each TDC count corresponds to 25/30 ns, whereas a unit of BX_COUNTER corresponds to 25 ns, and the ORBIT_CNT is increased every `x` BX_COUNTER. This allows to store the time in a similar way to hours, minutes and seconds.

In [11]:
# If haven't downloaded it yet, please get the data file with wget
!wget https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt -P ./data/

--2022-11-23 13:24:35--  https://www.dropbox.com/s/xvjzaxzz3ysphme/data_000637.txt
Risoluzione di www.dropbox.com (www.dropbox.com)... 162.125.69.18
Connessione a www.dropbox.com (www.dropbox.com)|162.125.69.18|:443... connesso.
Richiesta HTTP inviata, in attesa di risposta... 302 Found
Posizione: /s/raw/xvjzaxzz3ysphme/data_000637.txt [segue]
--2022-11-23 13:24:35--  https://www.dropbox.com/s/raw/xvjzaxzz3ysphme/data_000637.txt
Riutilizzo della connessione esistente a www.dropbox.com:443.
Richiesta HTTP inviata, in attesa di risposta... 302 Found
Posizione: https://ucf450284d8c4c7e66fe908da213.dl.dropboxusercontent.com/cd/0/inline/BxSuCBbAt4MUesgLELPpXobZFh06OvO3rEltw-s7dCccNyVrTGk_L5t6BWiKrOuOWtO1d9otDn-qAyC1In5-dah8driCiceuAMcRpQQAN0ypqgR9LiDhhrPtkxrWIKmtUhlxph8PW37p6-HomDznfx-ZmgbqdT5-nb3ggmV9pujVsg/file# [segue]
--2022-11-23 13:24:36--  https://ucf450284d8c4c7e66fe908da213.dl.dropboxusercontent.com/cd/0/inline/BxSuCBbAt4MUesgLELPpXobZFh06OvO3rEltw-s7dCccNyVrTGk_L5t6BWiKrOuOWtO1d9o

1\. Create a Pandas DataFrame reading N rows of the `data/data_000637.txt` dataset. Choose N to be smaller than or equal to the maximum number of rows and larger that 10k (check the documentation).

In [31]:
import numpy as np
import pandas as pd
import datetime as dt


dataf = pd.read_csv('data/data_000637.txt.2', nrows = np.random.randint(1000,1310721))
dataf

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS
0          1     0          123  3869200167        2374        26
1          1     0          124  3869200167        2374        27
2          1     0           63  3869200167        2553        28
3          1     0           64  3869200167        2558        19
4          1     0           64  3869200167        2760        25
...      ...   ...          ...         ...         ...       ...
877941     1     1            1  3869208195        2905        27
877942     1     0           62  3869208195        2903        19
877943     1     0           37  3869208195        2907        23
877944     1     1            2  3869208195        2910        13
877945     1     0          139  3869208195        2912         0

[877946 rows x 6 columns]

2\. Estimate the number of BX in a ORBIT (the value `x`).

*Hint*: check when the BX counter reaches the maximum value before being reset to 0.

In [32]:
x = dataf.groupby('ORBIT_CNT')['BX_COUNTER'].transform(np.max).unique().mean()
x

2559.707744107744

3\. Create a new column with the absolute time in ns (as a combination of the other three columns with timing information) since the beginning of the data acquisition, and convert the new column to a Time Series.

In [33]:
dataf['Absolute Time (ns)'] = (dataf['TDC_MEAS']*(25/30) + dataf['BX_COUNTER']*25).cumsum()
dataf['Absolute Time (Datetime)'] = pd.to_datetime(dataf['Absolute Time (ns)'])
dataf

HEAD  FPGA  TDC_CHANNEL   ORBIT_CNT  BX_COUNTER  TDC_MEAS  \
0          1     0          123  3869200167        2374        26   
1          1     0          124  3869200167        2374        27   
2          1     0           63  3869200167        2553        28   
3          1     0           64  3869200167        2558        19   
4          1     0           64  3869200167        2760        25   
...      ...   ...          ...         ...         ...       ...   
877941     1     1            1  3869208195        2905        27   
877942     1     0           62  3869208195        2903        19   
877943     1     0           37  3869208195        2907        23   
877944     1     1            2  3869208195        2910        13   
877945     1     0          139  3869208195        2912         0   

        Absolute Time (ns)      Absolute Time (Datetime)  
0             5.937167e+04 1970-01-01 00:00:00.000059371  
1             1.187442e+05 1970-01-01 00:00:00.000118744  
2             1.825925e+05 1970-01-01 00:00:00.000182592  
3             2.465583e+05 1970-01-01 00:00:00.000246558  
4             3.155792e+05 1970-01-01 00:00:00.000315579  
...                    ...                           ...  
877941        3.923131e+10 1970-01-01 00:00:39.231306982  
877942        3.923138e+10 1970-01-01 00:00:39.231379573  
877943        3.923145e+10 1970-01-01 00:00:39.231452267  
877944        3.923153e+10 1970-01-01 00:00:39.231525028  
877945        3.923160e+10 1970-01-01 00:00:39.231597828  

[877946 rows x 8 columns]

4\. Find out the duration of the data taking in hours, minutes and seconds, by using the features of the Time Series. Perform this check reading the whole dataset.

5\. Use the `.groupby()` method to find out the noisy channels, i.e. the TDC channels with most counts (print to screen the top 3 and the corresponding counts)

6\. Count the number of non-empty orbits (i.e. the number of orbits with at least one hit).

7\. Count the number of unique orbits with at least one measurement from TDC_CHANNEL=139.

8\. Create two Series (one for each FPGA) that have the TDC channel as index, and the number of counts for the corresponding TDC channel as values.

9\. **Optional:** Create two histograms (one for each FPGA) that show the number of counts for each TDC channel.